### PR TITLE
Add universal qr scanner

### LINKED ISSUE #10764: Support universal QR code scanner
# Problem
Follow up on PR #10729 that was fixed #10690 

This issue enabled scanning 1:1 chats, but did not yet included support for public chat or dapp scanning. A universal scanner can resolve this.

Universal scanner has existed in the past, but was removed from V1 in order to simplify the action list under the plus-button. At the time there was no design to give the Universal QR code scanner a more appropriate place

## Implementation
UI position for QR code scanner should follow [Figma design](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=375%3A3183)

Home Tab -_ New

### Chat

* [ ] If user scans public chat universal link like https://join.status.im/markets open public chat 
* [ ] If user scans universal links with status ENS name, open 1:1 chat 
* [ ] If user scans universal link with ENS name with other domain, open 1:1 chat 
* [ ] If user scans other user profile key, open 1:1 chat 
* [ ] If user scans universal link with own profile chat key, open dialog with copy Title: "That's you" Body: "To start a chat choose someone else"
* [ ] If user scans universal link with other user chat key, open 1:1 chat 

### Wallet

* [ ] If users scans wallet address open `Wallet` > `Send` > `To` field includes address

### dapps

* [ ] If user scans universal link with Dapp (https://join.status.im/b/ens.domains) open dapp in browser

### Error

* [ ] If user scans invalid values, dialog: "Unable to read this code"

## Acceptance Criteria
Above behavior

## Future Steps
Automatically open Send transaction doesn't always make sense when scanning an address. The user might actually scan to send a tx request in payment scenarios. This might need updating in the future. Currently, Send transaction is the most frequently used use case

### PR DESCRIPTION
**Pull request by Ferossgp**
Closes #10764

Added [bidi](https://github.com/juxt/bidi) to handle routes instead of regexp. Also unified the QR and Universal links, to use same routing as they use the same links.

### CODE REVIEW

**Comment by churik** [comment]
@Ferossgp can you pleae rebase it? thanks!

**Comment by churik** [comment]
@Ferossgp 
### ISSUE 1: on scanning chat keys or universal links app is redirected to profile view with profile of another user

And to see again your profile you need to go back.
Moreover, the profile of another user stays opened in "Profile" tab after performing actions (i.e. sending messages or blocking)

**Expected result:**
profile of another user will be opened in home view

**Actual result:**

BAACAgQAAxkBAAEqCG9fLCMfmEOTUFMmYJXJDVsu8jnD_AACUAwAAmQvYFFwptFxYH3GUxoE_1596728137 93


### ISSUE 2: unable to scan valid QR after scanning invalid ENS name
**Steps:**
- scan QR with invalid ENS (i.e. `https://join.status.im/u/tanyatest.eth`)
- try to scan QR with valid ENS (i.e. `https://join.status.im/u/tanyatest1.eth`)

**Expected result:**
can see error; can scan valid QR normally
**Actual result:**
no error; at attempt to scan valid QR - no redirect

### ISSUE 3: on scanning deep link with public chat - null is not an object error
**Steps:**
- scan QR with valid public chat name ddep link i.e. `https://join.status.im/markets`

**Expected result:**
joined `markets`

**Actual result:**
error
IMAGE 2020-08-06 17:57:24


### ISSUE 4: doesn't support universal links to dapps with full URLs
**Steps:**
- scan QR with valid link i.e. `https://join.status.im/b/https://www.google.com/`

**Expected result:**
open in Status

**Actual result:**
Unable to read the code

### ISSUE 5: error at attempt to scan EIP link 
**Steps:**
- scan QR with valid EIP link i.e. `ethereum:0x3d597789ea16054a084ac84ce87f50df9198f415?value=1e15`

**Expected result:**
Send transaction screen with 0.001 ETH with "To" 0x3d597789ea16054a084ac84ce87f50df9198f415

**Actual result:**
Unable to read the code
IMAGE 2020-08-06 18:12:08
At attempt to scan valid wallet address, got 'Unable to read this code', so seems it is not finished.

**Comment by Ferossgp** [comment]
@churik 
> ISSUE 3: on scanning deep link with public chat - null is not an object error

Cannot replicate, maybe there was some context before that action, or can you please share INFO logs?

All other issues fixed

**Comment by churik** [comment]
@Ferossgp thanks for fixes!
also can't reproduce issue 3.

About other issues:
### ISSUE 5: EIP links are still not fully supported (for tokens, ENS names, payment links)
More context of supported params is here: https://github.com/status-im/status-react/pull/10280

Links for testing:
- 20 ZRX to 0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8
`ethereum:pay-0xe41d2489571d322189246dafa5ebde1f4699f498/transfer?address=0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8&uint256=2e1`

- 0.001 ETH with "To" 0x3d597789ea16054a084ac84ce87f50df9198f415
`ethereum:0x3d597789ea16054a084ac84ce87f50df9198f415?value=1e15`

- donate 0.11 SNT to 3esmit.stateofus.eth
`ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=3esmit.stateofus.eth&uint256=1.1e-1`

- with chain_id
`ethereum:0x3d597789ea16054a084ac84ce87f50df9198f415@1?value=1e16`

Test case is here: https://ethstatus.testrail.net/index.php?/cases/view/6282

Also in case of scanning valid wallet address as it mentioned in https://github.com/status-im/status-react/issues/10764 should open If users scans wallet address open Wallet > Send > To field includes address.


### ISSUE 6: after scanning another user profile scanning your profile key redirects to another user profile
**Steps:**
- create multiaccount
- scan QR with another user profile chat key or ENS (i.e. `https://join.status.im/u/tanyatest1.eth`)
-  scan QR with own profile  chat key 

**Expected result:**
redirected to own profile

**Actual result:**
redirected to   tanyatest1.eth (opened in Profile tab, like in issue 1)

**Comment by Ferossgp** [comment]
@churik fixed

File: src/status_im/qr_scanner/core.cljs [hunk 1]
```diff
@@ -16,11 +22,73 @@
...
+                          :content    (i18n/label :t/ens-name-not-found)
+                          :on-dismiss #(re-frame/dispatch [:navigate-to :home])}})))
+
+(fx/defn handle-eip681 [cofx data]
+  (fx/merge cofx
+            {:dispatch [:wallet/parse-eip681-uri-and-resolve-ens data]}
+            (navigation/navigate-to-cofx :tabs {:screen :wallet})))
+
+(fx/defn match-scan
+  {:events [::match-scanned-value]}
```
Comment (flexsurfer) [inline src/status_im/qr_scanner/core.cljs:78]: could you please elaborate why handlers  are here and not in the router ns ?thanks

Comment (flexsurfer) [reply src/status_im/qr_scanner/core.cljs:78]: should it be handled different if we scan it or if we open universal link?

**Comment by churik** [comment]
@Ferossgp 
in the last version with commit https://github.com/status-im/status-react/commit/29a35bd058c413e199228645d7b65fdc6b1fc50f wallet scanner (wallet view > top right corner) stop working.
At attempt to scan valid EIP link (i.e. 'ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=3esmit.stateofus.eth&uint256=1.1e-1') got `<no value>` error

File: src/status_im/qr_scanner/core.cljs (hunk 1 above)
Comment (Ferossgp) [reply src/status_im/qr_scanner/core.cljs:78]: Handlers are almost similar, the only difference is on errors strings. Also in some places we do not show error in Universal Links, while in QR we say that we can't scan this code

**Comment by Ferossgp** [comment]
@churik fixed wallet scanner

**Review by flexsurfer — APPROVED** [review]

**Comment by churik** [comment]
Tested universal code scanner + existing QR code scanners from chat view, wallet view, bootnode, mailserver QR code scanner on IOS 13 and Android 10

Awesome work!

**Comment by churik** [comment]
@Ferossgp new issue with redirect to user profile from universal link:
when tapping on universal link and select "Open in status" another user profile view is opened on profile tab and there is no way to return on your own profile (like it was in issue 1 in  https://github.com/status-im/status-react/pull/11014#issuecomment-670027497, but button "Back" redirects to home view.

Thanks for the clue!

**Comment by churik** [comment]
Fixed and verified. Ready to be merged!

**Pull request merged**